### PR TITLE
Mix --max-fail opt

### PR DESCRIPTION
**Pull request by Kartstig**
- Relates to https://github.com/elixir-lang/elixir/issues/4586

### CODE REVIEW

File: lib/mix/lib/mix/tasks/test.ex [hunk 1]
```diff
@@ -277,6 +279,9 @@ defmodule Mix.Tasks.Test do
         option_only_present? = Keyword.has_key?(opts, :only)
 
         cond do
+          failures >= opts[:max_fail] ->
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/test.ex]: This is only invoked after all tests run, so it is not going to give you the desired effect.

File: lib/mix/test/mix/tasks/test_test.exs [hunk 2]
```diff
@@ -46,21 +52,21 @@ defmodule Mix.Tasks.TestTest do
...
       assert_stale_run_output("""
       No stale tests
       """)
-    end
+    end)
   end
 
   test "--stale: runs tests that depend on modified modules" do
-    in_fixture "test_stale", fn ->
+    in_fixture("test_stale", fn ->
```
Comment (josevalim) [inline lib/mix/test/mix/tasks/test_test.exs]: Whatever you are using to format the code is not considering the .formatter.exs at the root of the Elixir codebase, as it should not add those parens. :)

File: lib/mix/lib/mix/tasks/test.ex (hunk 1 above)
Comment (Kartstig) [reply lib/mix/lib/mix/tasks/test.ex]: Right, it will be caught and it won't stop until the suite is done. Should a call to `agent_stop` be made?

File: lib/mix/test/mix/tasks/test_test.exs (hunk 2 above)
Comment (Kartstig) [reply lib/mix/test/mix/tasks/test_test.exs]: Ah. Yeah it seems like it's coming from my Sublime Text formatter. I'll remove these changes

File: lib/mix/lib/mix/tasks/test.ex (hunk 1 above)
Comment (josevalim) [reply lib/mix/lib/mix/tasks/test.ex]: @Kartstig we would need to add support for this in ExUnit instead.

File: lib/mix/lib/mix/tasks/test.ex [hunk 3]
```diff
@@ -58,6 +58,7 @@ defmodule Mix.Tasks.Test do
     * `--listen-on-stdin` - runs tests, and then listens on stdin. Receiving a newline will
       result in the tests being run again. Very useful when combined with `--stale` and
       external commands which produce output on stdout upon file system modification.
+    * `--max-fail` - sets the maximum number of tests to allow to fail before stopping
```
Comment (eksperimental) [inline lib/mix/lib/mix/tasks/test.ex]: You are missing the trailing period at the end of this sentence

File: lib/mix/test/mix/tasks/test_test.exs [hunk 4]
```diff
@@ -46,11 +52,11 @@ defmodule Mix.Tasks.TestTest do
   defp ex_unit_opts_from_given(passed) do
     passed
     |> ex_unit_opts()
-    |> Keyword.drop([:failures_manifest_file, :autorun])
+    |> Keyword.drop(@default_opts)
   end
 
   test "--stale: runs all tests for first run, then none on second" do
-    in_fixture "test_stale", fn ->
+    in_fixture"test_stale", fn ->
```
Comment (eksperimental) [inline lib/mix/test/mix/tasks/test_test.exs]: Is the removal of the white space intentional?

File: lib/mix/lib/mix/tasks/test.ex [hunk 5]
```diff
@@ -450,6 +457,16 @@ defmodule Mix.Tasks.Test do
     Enum.filter(matched_test_files, &MapSet.member?(allowed_files, Path.expand(&1)))
   end
 
+  defp max_fail_opts(opts) do
+    case Keyword.fetch(opts, :max_fail) do
+      {:ok, _} ->
+        opts
```
Comment (eksperimental) [inline lib/mix/lib/mix/tasks/test.ex]: What happens when max fail is a negative integer? I think we should raise.

File: lib/mix/test/mix/tasks/test_test.exs [hunk 6]
```diff
@@ -3,10 +3,16 @@ Code.require_file("../../test_helper.exs", __DIR__)
...
+  @default_opts [:failures_manifest_file, :autorun, :max_fail]
+
   test "ex_unit_opts/1 returns ex unit options" do
     assert ex_unit_opts_from_given(unknown: "ok", seed: 13) == [seed: 13]
   end
 
+  test "ex_unit_opts/1 returns max_fail" do
+    assert ex_unit_opts_from_given(max_fail: 13) == [max_fail: 13]
+  end
+
```
Comment (eksperimental) [inline lib/mix/test/mix/tasks/test_test.exs]: It would be good to have a test for negative integers

File: lib/mix/lib/mix/tasks/test.ex (hunk 1 above)
Comment (eksperimental) [reply lib/mix/lib/mix/tasks/test.ex]: If max fail is zero, this will fail eventhough no failures happened. So we need to cover this corner case

File: lib/mix/lib/mix/tasks/test.ex (hunk 3 above)
Comment (eksperimental) [reply lib/mix/lib/mix/tasks/test.ex]: Actually the logic is wrong. It should be `>`

File: lib/mix/lib/mix/tasks/test.ex (hunk 1 above)
Comment (eksperimental) [reply lib/mix/lib/mix/tasks/test.ex]: Actually the logic is wrong. It should be `>`

**Comment by josevalim** [comment]
> I originally branched off v1.6.3, but then when trying to merge the latest master I had a ton of conflicts, so I cherry-picked my commits off the latest master, but now I get the following errors:

Run `make clean` and start again. :)

File: lib/mix/test/mix/tasks/test_test.exs (hunk 4 above)
Comment (Kartstig) [reply lib/mix/test/mix/tasks/test_test.exs]: Not intentional

**Comment by eksperimental** [comment]
I can take care of finishing this feature but for some reason I thought something had to be decided on the Elixirs core side, or I may be mistaken with another PR

**Comment by josevalim** [comment]
I don't think there is nothing pending on our side. We should call it `--max-failures` and implement it rather on the ExUnit.Runner side. The other thing is what to do with concurrent tests. When a test fails other N tests may be running and we can't just call them, otherwise we would skip cleanup.

**Comment by Kartstig** [comment]
Thanks @eksperimental. I got stuck on this after Jose pointed out that the actual feature needs to be integrated into ExUnit. I couldn't find the right place for implementation and lost track of this. If you could point me in the right direction, I may be able to contribute
